### PR TITLE
Add Foundry and custom durable history samples

### PR DESCRIPTION
**Pull request by tamirdresher**
## Summary

Adds the experimental durable history Samples 08 and 09. This is stack position **5 of 6**, based on `tamirdresher-microsoft-stack-4-retention-metrics`; Sample 10 is intentionally excluded.

These are draft/experimental samples only. They do not expose a public `EnableMailboxWrites` API or activate production schema2 behavior. Samples and tests use an internal, default-disabled activation path. Mixed-runtime production use remains blocked until Python, dashboard, and poller compatibility reaches the rollout floor.

- `08_FoundryManagedAgent` uses a stable logical service key with `KeepAll`. It exercises an actual versioned Microsoft Foundry durable proxy and cold restart, preserves mailbox and completion state without mirroring the service transcript into entity `conversationHistory`, restores the service binding and opaque continuation/reference, verifies continued callability, and cleans up the exact agent version created by the sample.
- `09_CustomHistoryProvider` uses a stable external provider key with `KeepAll`. It verifies proxy start and cold-restart reference continuity/callability with a compatible provider, rejects a changed provider key, preserves mailbox and completion state without transcript mirrors, projects bounded input into model context, and cleans up its local history. Marker input is explicitly limited to 1 KiB of UTF-8 before durable activity starts. Its cumulative history test exceeds 1 MiB through moderate records and does not claim to solve a single oversized payload.

## Configuration

The samples use:

- `FOUNDRY_PROJECT_ENDPOINT`
- `FOUNDRY_MODEL`
- `DURABLE_TASK_SCHEDULER_CONNECTION_STRING`

Authentication uses `DefaultAzureCredential`; no credentials or connection secrets are included.

## Validation

- Sample 09 tests: **13/13 passed**.
- Registration/Foundry tests: **4/4 passed**.
- DurableTask tests: **1338/1338 passed**.
- Hosting tests: **223/223 passed**.
- Sample 08 and Sample 09 Release builds passed.
- Full Release build: **38 projects, 0 warnings, 0 errors**.
- Package and symbol-package creation passed.
- Formatting, analyzers, and scope checks passed.
- Sample verifier: **0 run, 2 secure checks skipped** because required live-service configuration was unavailable.
- Final independent `gpt-5.6-sol` review: **ACCEPT** for exact head `d3c09e9ad91602a596993950a8b893a8756d26a8`.

## Known limitations

- The custom provider is intentionally text-focused and backed by local JSON.
- External provider writes and durable entity commits are not transactionally coupled.
- The samples do not provide payload offload, distributed storage, exactly-once behavior, or a solution for one oversized request/tool result.
- No production-readiness or production schema2 activation is claimed.

## Change summary

- Expected stack base commit: `1479c5322f61544fd13be65805e6bf0b96d668b3`
- Head commit: `d3c09e9ad91602a596993950a8b893a8756d26a8`
- Incremental diff: **17 files changed, 2,139 insertions**
- Cumulative diff from repository `origin/main` (`62afdbac03f0a81b0917abe6328203b706a2f294`): **130 files changed, 27,074 insertions, 395 deletions**

### Incremental commits

1. `04380e51e964a8330304e0645a3b8d2cfe1c881c` Add gated durable history samples
2. `d3c09e9ad91602a596993950a8b893a8756d26a8` Bound custom history sample marker input

### CODE REVIEW

File: dotnet/samples/DurableAgents/ConsoleApps/09_CustomHistoryProvider/JsonFileChatHistoryProvider.cs [hunk 1]
```diff
@@ -0,0 +1,413 @@
...
+            throw new InvalidOperationException("A session is required for file-backed history.");
+        HistoryReference reference = this.GetOrCreateReference(session);
+
+        await this._gate.WaitAsync(cancellationToken);
+        try
+        {
+            List<StoredChatMessage> stored =
+                await this.ReadStoredMessagesAsync(reference, cancellationToken);
+            stored.AddRange(context.RequestMessages.Select(ToStoredMessage));
+            stored.AddRange((context.ResponseMessages ?? []).Select(ToStoredMessage));
```
[the remark this answers is not in the document]
Comment (tamirdresher) [reply dotnet/samples/DurableAgents/ConsoleApps/09_CustomHistoryProvider/JsonFileChatHistoryProvider.cs:207]: The second-invocation regression test passes the provider-expanded input into InvokedContext and asserts persisted count grows by exactly two, not by the history prefix. The pinned callback exposes the current request suffix here; all 15 provider tests pass. No deduplication heuristic is needed.

File: dotnet/samples/DurableAgents/ConsoleApps/09_CustomHistoryProvider/JsonFileChatHistoryProvider.cs [hunk 2]
```diff
@@ -0,0 +1,413 @@
...
+    }
+
+    private async Task<List<StoredChatMessage>> ReadStoredMessagesAsync(
+        HistoryReference reference,
+        CancellationToken cancellationToken)
+    {
+        string path = this.GetHistoryPath(reference);
+        if (!File.Exists(path))
+        {
+            return [];
```
[the remark this answers is not in the document]
Comment (tamirdresher) [reply dotnet/samples/DurableAgents/ConsoleApps/09_CustomHistoryProvider/JsonFileChatHistoryProvider.cs]: Fixed. The provider now tracks references allocated by the current instance; only those may have no file before first write. A restored continuation pointing to missing storage throws FileNotFoundException before supplying empty history. A restart regression test covers this fail-closed behavior.

File: dotnet/samples/DurableAgents/ConsoleApps/09_CustomHistoryProvider/JsonFileChatHistoryProvider.cs [hunk 3]
```diff
@@ -0,0 +1,430 @@
...
+                    "The external chat history for the restored session is unavailable.",
+                    path);
+        }
+
+        await using FileStream stream = File.OpenRead(path);
+        return await JsonSerializer.DeserializeAsync<List<StoredChatMessage>>(
+                stream,
+                s_jsonOptions,
+                cancellationToken)
+            ?? [];
```
[the remark this answers is not in the document]
Comment (tamirdresher) [reply dotnet/samples/DurableAgents/ConsoleApps/09_CustomHistoryProvider/JsonFileChatHistoryProvider.cs:329]: Agreed. This is a sample-provider corruption edge case rather than a core runtime contract change, so the one-line fail-closed behavior and regression test are tracked as follow-up #105 to avoid another full stack rewrite.
